In [1]:
import pandas as pd
helper = pd.read_csv(r'C:\Users\c3205\Documents\Code\python\draw\draws\V6\helper.csv', dtype=str, encoding_errors='replace')

In [2]:
# Find all rows where any column contains the replacement character '�' (often used for unknown symbols)
mask = helper.apply(lambda row: row.astype(str).str.contains('�').any(), axis=1)
helper[mask]

,DAY,DATE,TIME,FIELD,ROUND,GRADE,TEAM 1,v,TEAM 2
413,FRI,20-Jun-25,19:00,EF,11,PHL,Tigers�,v,Wests�
415,SUN,22-Jun-25,8:30,EF,11,3rd,Port Stephens�,v,Wests Red
416,SUN,22-Jun-25,10:00,EF,11,2nd,Wests,v,Tigers�
417,SUN,22-Jun-25,11:30,EF,11,3rd,University�,v,Wests Green�
418,SUN,22-Jun-25,13:00,EF,11,5th,Colts�,v,University Seapigs�
419,SUN,22-Jun-25,14:30,EF,11,6th,Wests�,v,Colts�
420,SUN,22-Jun-25,16:00,EF,11,5th,Tigers Yellow�,v,Wests�
422,SUN,22-Jun-25,8:30,WF,11,4th,Norths�,v,Tigers�
423,SUN,22-Jun-25,10:00,WF,11,6th,Souths�,v,University Gentlemen�
424,SUN,22-Jun-25,11:30,WF,11,3rd,Souths�,v,Norths


In [3]:
helper = helper[helper['DATE'] != '27-April-2025']

In [3]:
helper['TEAM 1'] = helper['TEAM 1'].astype(str)
helper['TEAM 2'] = helper['TEAM 2'].astype(str)

helper['TEAM 1'] = helper['TEAM 1'].str.strip()
helper['TEAM 2'] = helper['TEAM 2'].str.strip()
helper['GRADE'] = helper['GRADE'].str.strip()

helper['TEAM 1'] = helper['TEAM 1'].str.replace('�', '')
helper['TEAM 2'] = helper['TEAM 2'].str.replace('�', '')
helper['GRADE'] = helper['GRADE'].str.replace('�', '')



helper['TEAM 1'] = helper['TEAM 1'].str.replace('University Gentleman', 'University Gentlemen')
helper['TEAM 2'] = helper['TEAM 2'].str.replace('University Gentleman', 'University Gentlemen')

helper['TEAM 1'] = helper['TEAM 1'].str.replace('Unversity Seapigs', 'University Seapigs')
helper['TEAM 2'] = helper['TEAM 2'].str.replace('Unversity Seapigs', 'University Seapigs')

helper['TEAM 1'] = helper['TEAM 1'].str.replace('Unversity', 'University')
helper['TEAM 2'] = helper['TEAM 2'].str.replace('Unversity', 'University')

helper['TEAM 1'] = helper['TEAM 1'].str.replace('Universiity', 'University')
helper['TEAM 2'] = helper['TEAM 2'].str.replace('Universiity', 'University')

helper['TEAM 1'] = helper['TEAM 1'].str.replace('West Red', 'Wests Red')
helper['TEAM 2'] = helper['TEAM 2'].str.replace('West Red', 'Wests Red')

helper['TEAM 1'] = helper['TEAM 1'].str.replace('West Green', 'Wests Green')
helper['TEAM 2'] = helper['TEAM 2'].str.replace('West Green', 'Wests Green')


In [24]:
helper['TEAM 2'].unique()

array(['Wests Green', 'University', 'Wests', 'Wests Red', 'Souths', 'nan',
       'University Seapigs', 'Maitland', 'Port Stephens', 'Tigers Yellow',
       'Tigers', 'University Gentlemen', 'Norths', 'BYE', 'TEAM 2',
       'Gosford', 'Tigers Black', 'Crusaders', 'EASTER', 'Colts'],
      dtype=object)

In [10]:
import pandas as pd

# Remove unwanted team labels
cleaned = helper[
    ~helper['TEAM 1'].isin(['EASTER', 'TEAM 1', 'TEAM 2']) &
    ~helper['TEAM 2'].isin(['EASTER', 'TEAM 1', 'TEAM 2'])
].copy()

# Create a standardised pairing regardless of order
cleaned['Team_A'] = cleaned[['TEAM 1', 'TEAM 2']].min(axis=1)
cleaned['Team_B'] = cleaned[['TEAM 1', 'TEAM 2']].max(axis=1)

# Now group by GRADE and the unordered pair, and count games
match_counts = (
    cleaned
    .groupby(['GRADE', 'Team_A', 'Team_B'])
    .size()
    .reset_index(name='Matches')
)

# Now for each GRADE, pivot into a symmetric table
grades = match_counts['GRADE'].unique()
for grade in grades:
    # if grade == 'PHL':
        sub = match_counts[match_counts['GRADE'] == grade]

        # Get full list of teams involved in this grade
        teams = pd.unique(sub[['Team_A', 'Team_B']].values.ravel())

        # Create full symmetric index and columns
        pivot = sub.pivot(index='Team_A', columns='Team_B', values='Matches').fillna(0).astype(int)
        pivot = pivot.reindex(index=teams, columns=teams, fill_value=0)

        # Symmetrize the matrix by copying upper + lower triangle
        symmetrical = pivot.add(pivot.T, fill_value=0).astype(int)

        # Compute total games played by team
        symmetrical['Total'] = symmetrical.sum(axis=1)
        symmetrical.loc['Total'] = symmetrical.sum(axis=0)

        print(f"\nGRADE: {grade}")
        print(symmetrical)



GRADE: 2nd
Team_B    BYE  Maitland  Norths  Souths  Tigers  Wests  Total
Team_A                                                       
BYE         0         5       4       4       4      5     22
Maitland    5         0       4       3       4      4     20
Norths      4         4       0       4       4      4     20
Souths      4         3       4       0       4      4     19
Tigers      4         4       4       4       0      3     19
Wests       5         4       4       4       3      0     20
Total      22        20      20      19      19     20    120

GRADE: 3rd
Team_B              BYE  Crusaders  Norths  Port Stephens  Souths  Tigers  \
Team_A                                                                      
BYE                   0          2       4              2       1       1   
Crusaders             2          0       2              2       2       2   
Norths                4          2       0              2       3       2   
Port Stephens         2          

In [ ]:
# Function to check home games for a team
def check_home_games(team, home_ground):
    team_games = cleaned[
        ((cleaned['TEAM 1'] == team) & (cleaned['FIELD'] == home_ground)) |
        ((cleaned['TEAM 2'] == team) & (cleaned['FIELD'] == home_ground))
    ]
    
    # Group by grade and opponent
    home_game_counts = team_games.groupby(['GRADE', 'Team_A', 'Team_B']).size().reset_index(name='Home_Games')
    
    # Filter for games involving the team
    home_game_counts = home_game_counts[
        (home_game_counts['Team_A'] == team) | (home_game_counts['Team_B'] == team)
    ]
    
    # Check if home games are exactly half of total games
    for _, row in home_game_counts.iterrows():
        grade = row['GRADE']
        opponent = row['Team_B'] if row['Team_A'] == team else row['Team_A']
        total_games = match_counts[
            (match_counts['GRADE'] == grade) &
            (((match_counts['Team_A'] == team) & (match_counts['Team_B'] == opponent)) |
             ((match_counts['Team_B'] == team) & (match_counts['Team_A'] == opponent)))
        ]['Matches'].values[0]
        
        if row['Home_Games'] != total_games / 2:
            print(f"Discrepancy found: {team} vs {opponent} in grade {grade} - Home games: {row['Home_Games']}, Total games: {total_games}")

# Check for Maitland
check_home_games('Maitland', 'MAI')

# Check for Gosford
check_home_games('Gosford', 'CC')

In [28]:
first_team = 'Wests Green'
second_team = 'Maitland'
grade = '4th'
# helper[(((helper['TEAM 1'] == first_team) & (helper['TEAM 2'] == second_team) ) | ((helper['TEAM 2'] == first_team) & (helper['TEAM 1'] == second_team) )) & (helper['GRADE'] ==  grade)]
helper[(((helper['TEAM 1'] == first_team)  ) | ((helper['TEAM 2'] == first_team) )) & (helper['GRADE'] ==  grade)]


,DAY,DATE,TIME,FIELD,ROUND,GRADE,TEAM 1,v,TEAM 2
10,SUN,23-March-2025,10:00,WF,1,4th,Norths,v,Wests Green
35,SUN,30-March-2025,16:30,EF,2,4th,University,NaN,Wests Green
85,SUN,06-April-2025,NaN,NaN,3,4th,Wests Green,NaN,BYE
113,SUN,13-April-2025,14:30,SF,4,4th,Maitland,v,Wests Green
166,SUN,27-April-2025,10:30,WF,5,4th,Colts,v,Wests Green
226,SUN,11-May-2025,13:30,WF,NaN,4th,Souths,v,Wests Green
277,SUN,18-May-2025,NaN,NaN,7,4th,Wests Green,NaN,BYE
291,SUN,25-May-2025,11:30,WF,8,4th,Tigers,v,Wests Green
318,SUN,01-June-2025,16:00,EF,9,4th,Port Stephens,v,Wests Green
383,SUN,15-June-2025,10:00,EF,10,4th,Wests Red,v,Wests Green


In [ ]:
first_team = 'Souths'
second_team = 'Maitland'
grade = '2nd'
helper[(((helper['TEAM 1'] == first_team) & (helper['TEAM 2'] == second_team) ) | ((helper['TEAM 2'] == first_team) & (helper['TEAM 1'] == second_team) )) & (helper['GRADE'] ==  grade)]
# helper[(((helper['TEAM 1'] == first_team)  ) | ((helper['TEAM 2'] == first_team) )) & (helper['GRADE'] ==  grade)]


In [4]:
# ...existing code...
# Count unordered pair occurrences per grade and print ordered by grade
# (uses existing `match_counts` if present, otherwise compute it from `helper`)

# temp = helper.drop_duplicates(subset=['GRADE', 'TEAM 1', 'TEAM 2', 'ROUND']).copy()
temp = helper.copy()

temp = temp[
    ~temp['TEAM 1'].isin(['EASTER', 'TEAM 1', 'TEAM 2']) &
    ~temp['TEAM 2'].isin(['EASTER', 'TEAM 1', 'TEAM 2'])
].copy()
temp['Team_A'] = temp[['TEAM 1', 'TEAM 2']].min(axis=1)
temp['Team_B'] = temp[['TEAM 1', 'TEAM 2']].max(axis=1)
match_counts = temp.groupby(['GRADE', 'Team_A', 'Team_B']).size().reset_index(name='Matches')

total_total = 0
for grade in sorted(match_counts['GRADE'].dropna().unique()):
    print(f"GRADE: {grade}")
    sub = match_counts[match_counts['GRADE'] == grade].sort_values(['Matches', 'Team_A', 'Team_B'], ascending=[False, True, True])

    # running total of pairings (counts Matches as we iterate)
    running_total = 0
    for _, row in sub.iterrows():
        if row['Team_A'] != 'BYE':
            running_total += int(row['Matches'])
            print(f"  {row['Team_A']} vs {row['Team_B']}: {row['Matches']} (running total: {running_total})")
    total_total += running_total
    print(f"Total pairings in grade {grade}: {running_total}")
print(f"Total games across all divisions is: {total_total}")
# ...existing

GRADE: 2nd
  Maitland vs Norths: 4 (running total: 4)
  Maitland vs Souths: 4 (running total: 8)
  Maitland vs Tigers: 4 (running total: 12)
  Maitland vs Wests: 4 (running total: 16)
  Norths vs Souths: 4 (running total: 20)
  Norths vs Tigers: 4 (running total: 24)
  Norths vs Wests: 4 (running total: 28)
  Souths vs Tigers: 4 (running total: 32)
  Souths vs Wests: 4 (running total: 36)
  Tigers vs Wests: 4 (running total: 40)
Total pairings in grade 2nd: 40
GRADE: 3rd
  Crusaders vs Wests Red: 4 (running total: 4)
  Norths vs Souths: 3 (running total: 7)
  Port Stephens vs Tigers: 3 (running total: 10)
  Port Stephens vs University: 3 (running total: 13)
  Souths vs University Seapigs: 3 (running total: 16)
  Tigers vs Wests Green: 3 (running total: 19)
  University vs Wests Green: 3 (running total: 22)
  University vs Wests Red: 3 (running total: 25)
  Crusaders vs Norths: 2 (running total: 27)
  Crusaders vs Port Stephens: 2 (running total: 29)
  Crusaders vs Souths: 2 (running to

In [ ]:
+12 -22 to get -10 which is 419 games played

Expected Games per team
2nd 16 
3rd 16
4th 16
5th 18
6th 21
PHL 20

total games
2nd 40 
3rd 72
4th 72
5th 90
6th 84
PHL 60
sum 418

Expected Games per team (john addition)
2nd 16 
3rd 18
4th 18
5th 18
6th 21
PHL 20

total games
2nd 40 
3rd 81
4th 81
5th 90
6th 84
PHL 60
sum 436

plus one random for Maitland in thingo gives 437